mountpoint-s3-client/src/mock_client.rs

```diff
@@ -720,11 +731,25 @@ fn validate_checksum(
 #[derive(Clone, Debug)]
 pub struct MockBackpressureHandle {
     read_window_end_offset: Arc<AtomicU64>,
+    request_range: Range<u64>,
+    part_size: u64,
+    read_window_increment_failed: Arc<AtomicBool>,
 }
 
 impl ClientBackpressureHandle for MockBackpressureHandle {
     fn increment_read_window(&mut self, len: usize) {
-        self.read_window_end_offset.fetch_add(len as u64, Ordering::SeqCst);
+        let prev_read_window_end_offset = self.read_window_end_offset.fetch_add(len as u64, Ordering::SeqCst);
+        let read_window_end_offset = prev_read_window_end_offset + len as u64;
+        let relative_read_window_end = read_window_end_offset - self.request_range.start;
+        if read_window_end_offset < self.request_range.end && !relative_read_window_end.is_multiple_of(self.part_size) {
+            tracing::warn!(
```

Comment:
Are we always logging this warning? I.e. regardless of the `fail_on_non_aligned_read_window` setting?